rust-bot/src/main.rs

```diff
@@ -1,41 +1,108 @@
 use std::env;
 
 use serenity::{
-    model::{channel::Message, gateway::Ready},
+    framework::standard::macros::{check, command, group},
+    framework::standard::{Args, CheckResult, CommandOptions, CommandResult, StandardFramework},
+    model::{
+        channel::{
+            ChannelType, GuildChannel, Message, PermissionOverwrite, PermissionOverwriteType,
+        },
+        guild::Role,
+        id::RoleId,
+        permissions::Permissions,
+    },
     prelude::*,
+    Error,
 };
 
-struct Handler;
+#[group]
+#[commands(ping)]
+struct General;
 
-impl EventHandler for Handler {
-    // Set a handler for the `message` event - so that whenever a new message
-    // is received - the closure (or function) passed will be called.
-    //
-    // Event handlers are dispatched through a threadpool, and so multiple
-    // events can be dispatched simultaneously.
-    fn message(&self, ctx: Context, msg: Message) {
-        if msg.content == "!ping-corgo-rust" {
-            // Sending a message can fail, due to a network error, an
-            // authentication error, or lack of permissions to post in the
-            // channel, so log to stdout when some error happens, with a
-            // description of it.
-            if let Err(why) = msg.channel_id.say(&ctx.http, "Pong from Rust!") {
-                println!("Error sending message: {:?}", why);
-            }
-        }
-    }
+#[group]
+#[checks(Mod)]
+#[commands(create_cohort)]
+struct Mod;
 
-    // Set a handler to be called on the `ready` event. This is called when a
-    // shard is booted, and a READY payload is sent by Discord. This payload
-    // contains data like the current user's guild Ids, current user data,
-    // private channels, and more.
-    //
-    // In this case, just print what the current user's username is.
-    fn ready(&self, _: Context, ready: Ready) {
-        println!("{} is connected!", ready.user.name);
+#[command]
+fn ping(ctx: &mut Context, msg: &Message) -> CommandResult {
+    if let Err(why) = msg.channel_id.say(&ctx.http, "Pong!") {
+        println!("Error sending message: {}", why);
     }
+    return Ok(());
+}
+
+// This command provisions out a channel with permissions
+// to read/write for users with the corresponding role. It
+// will also make the channel read-only for other users.
+#[command]
+fn create_cohort(ctx: &mut Context, msg: &Message, mut args: Args) -> CommandResult {
+    msg.channel_id
+        .say(&ctx.http, "Spinning up Adventure Club...")?;
+
+    let (cohort_name, channel_name) = gen_names(args.single::<String>().unwrap());
+    let role = create_role(ctx, msg, &cohort_name);
+    let channel = create_channel(ctx, msg, &channel_name, role);
+
+    match channel {
+        Ok(channel) => {
+            let reply_msg = format!(
+                "Successfully created {}! Feel free to add users.",
+                channel.name
+            );
+            msg.reply(&ctx.http, reply_msg).unwrap();
```

Comment:
these unwraps should log data into honeycomb when @mdarrik gets the implementation up